### ISSUE TITLE
More scenarios for nullability attributes

### ISSUE THREAD

**Issue by jcouv** [dotnet/roslyn#36242]
More scenarios involving return value (search for uses of `ReturnTypeWithAnnotations` in `NullableWalker`):
- Test increment operator with different nullability on input and output, as well as attribute on output (VisitIncrementOperator)
- GetEnumerator with attribute on output (VisitForEachExpression)
- Current property with attribute on output (VisitForEachIterationVariables)
- Unary operator (VisitUnaryOperator, lifted and unlifted scenarios)
- Logical operator (InferResultNullability)
- GetAwaiter with attribute on output

Some other scenarios:
- `[DisallowNull]` on input of a conversion method
- `[AllowNull]` on input of a conversion method

Relates to https://github.com/dotnet/roslyn/issues/35816

**Comment by jcouv** [comment]
The lack of support for nullability attributes on unary operators was reported in https://devdiv.visualstudio.com/DevDiv/_workitems/edit/1696955
Copying the provided example below for reference:
```
using System.Diagnostics.CodeAnalysis;

namespace ConsoleApp1
{
    public class Test
    {
        public void TestFunc() { }

        [return: NotNullIfNotNull(nameof(test))]
        public static Test? TestFunc(Test? test) => test is null ? null : new();

        [return: NotNullIfNotNull(nameof(test))]
        public static Test? operator ~(Test? test) => test is null ? null : new();

        [return: NotNullIfNotNull(nameof(test))]
        public static Test? operator !(Test? test) => test is null ? null : new();

        [return: NotNullIfNotNull(nameof(test))]
        public static Test? operator +(Test? test) => test is null ? null : new();

        [return: NotNullIfNotNull(nameof(test))]
        public static Test? operator -(Test? test) => test is null ? null : new();

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator &(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator |(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator ^(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator +(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator -(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator *(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator /(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator %(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator >(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator  left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator >=(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator <=(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator ==(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        [return: NotNullIfNotNull(nameof(left))]
        [return: NotNullIfNotNull(nameof(right))]
        public static Test? operator !=(Test? left, Test? right) => left is null ? right : (right is null ? left : new());

        public override bool Equals(object? obj) => obj is Test;
        public override int GetHashCode() => base.GetHashCode();
    }

    internal class Program
    {
        static void Main(string[] args)
        {
            Test test = new();
            var test1 = Test.TestFunc(test);
            test1.TestFunc();
            var test2 = ~test;
            test2.TestFunc();
            var test3 = !test;
            test3.TestFunc();
            var test4 = +test;
            test4.TestFunc();
            var test5 = -test;
            test5.TestFunc();
            var test6 = test & test1;
            test6.TestFunc();
            var test7 = test | test1;
            test7.TestFunc();
            var test8 = test ^ test1;
            test8.TestFunc();
            var test9 = test + test1;
            test9.TestFunc();
            var test10 = test - test1;
            test10.TestFunc();
            var test11 = test * test1;
            test11.TestFunc();
            var test12 = test / test1;
            test12.TestFunc();
            var test13 = test % test1;
            test13.TestFunc();
            var test14 = test > test1;
            test14.TestFunc();
            var test15 = test < test1;
            test15.TestFunc();
            var test16 = test >= test1;
            test16.TestFunc();
            var test17 = test <= test1;
            test17.TestFunc();
            var test18 = test == test1;
            test18.TestFunc();
            var test19 = test != test1;
            test19.TestFunc();
        }
    }
}
```
